### PR TITLE
Issue #2594 - fix to SaveOptimizeTest failures due to timing issues

### LINKED ISSUE #2594: SaveOptimizeTest occasional failures on windows
getting failures in the following tests maybe about 50% of the time when I run them locally, usually only one or two of them fail at a time

`SaveOptimizeTest.testChangeMaxInactiveWithSaveOptimisation()`
`SaveOptimizeTest.testCleanSessionAfterSavePeriod()`
`SaveOptimizeTest.testDirtySession()`

https://github.com/eclipse/jetty.project/blob/154298af643a898601e374f1d7078bb33c5b40e4/tests/test-sessions/test-sessions-common/src/test/java/org/eclipse/jetty/server/session/SaveOptimizeTest.java#L481-L484

the failures occur on the assert checking when the SessionData was last saved (eg line 484)

```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at org.eclipse.jetty.server.session.SaveOptimizeTest.testChangeMaxInactiveWithSaveOptimisation(SaveOptimizeTest.java:484)
        ...
```

### PR DESCRIPTION
**Pull request by lachlan-roberts**
**Fix failing tests in SaveOptimizeTest due to timing issues** - closes #2594
Tests in SaveOptimizeTest were checking if the last time the SessionData had been saved was greater than it was before
if the second save occured on the same millisecond as the first it failed the assertion
now counting how many times the SessionData has been saved instead of when it was saved

Added memory barriers to TestContextScopeListener and TestSessionDataStore as they are being accessed from multiple different threads in the test

### CODE REVIEW

**Comment by gregw** [comment]
@lachlan-roberts you appear to have broken the UnixSocketTest on unix??

I don't think UnixSockets can work on windows??? The clue is in the name!

**Comment by joakime** [comment]
@gregw depends on how you execute on windows.
I know the wls layer can support it.
Even the powershell layer can be setup/configured to support it.
But I don't like that kind of requirement.

That's why the UnixSocketTest on the junit5 branch is setup to only execute on LINUX, MAC.
https://github.com/eclipse/jetty.project/blob/c86ccffcc9160607897b7e2216db485d6473494e/jetty-unixsocket/src/test/java/org/eclipse/jetty/unixsocket/UnixSocketTest.java#L54-L55

File: jetty-unixsocket/src/test/java/org/eclipse/jetty/unixsocket/UnixSocketTest.java [hunk 1]
```diff
@@ -49,20 +47,18 @@
...
+    File sockFile;
 
     @Before
     public void before() throws Exception
     {
         server = null;
         httpClient = null;
-        sockFile = Files.createTempFile(new File("/tmp").toPath(), "unix", ".sock" );
-        Files.deleteIfExists(sockFile);
+        sockFile = File.createTempFile("unix", ".sock" );
```
Comment (joakime) [inline jetty-unixsocket/src/test/java/org/eclipse/jetty/unixsocket/UnixSocketTest.java]: There should only be an `assumeFalse(OS.WINDOWS);` here in the `@Before`, all of the rest of the changes to UnixSocketTest are irrelevant.

**Comment by lachlan-roberts** [comment]
@gregw @joakime 
i believe the reason it was failing on linux was because i changed what it was doing
its failing without the `Files.deleteIfExists(sockFile)` but with that it works fine

File: jetty-unixsocket/src/test/java/org/eclipse/jetty/unixsocket/UnixSocketTest.java (hunk 1 above)
Comment (joakime) [reply jetty-unixsocket/src/test/java/org/eclipse/jetty/unixsocket/UnixSocketTest.java]: If you want to keep this, make sure the temp file is created within the `${maven.basedir}/target/` directory, that way you don't have to do any of the deleteIfExists logic to cleanup.

Use one of the MavenTestingUtils.getTarget*() methods that best suits your needs.

**Comment by lachlan-roberts** [comment]
was going to revert UnixSocketTest and only put `asumeTrue(OS.IS_UNIX)` in the `@Before` section however I noticed I was getting an exception coming up about 30% of the times I run the test, with the test still passing
```
2018-05-31 21:58:18.324:INFO::main: Logging initialized @439ms to org.eclipse.jetty.util.log.StdErrLog
2018-05-31 21:58:19.073:INFO:oejs.Server:main: jetty-9.4.z-SNAPSHOT; built: unknown; git: unknown; jvm 1.8.0_45-b14
2018-05-31 21:58:19.166:INFO:oejs.AbstractConnector:main: Started UnixSocketConnector@de3a06f{HTTP/1.1,[http/1.1]}{/tmp/unix4830866990252150602.sock}
2018-05-31 21:58:19.172:INFO:oejs.Server:main: Started @1287ms
2018-05-31 21:58:19.255:INFO:oeju.UnixSocketTest:qtp490150701-22: UnixSocketTest: request received
2018-05-31 21:58:19.298:WARN:oeji.ChannelEndPoint:HttpClient@76b10754-28: Ignoring key update for UnixSocketEndPoint@4e59e5ac{null<->null,CLOSED,fill=FI,flush=-,to=8/0}{io=1/1,kio=-1,kro=-1}->HttpConnectionOverHTTP@10e7e5fe(l:null <-> r:null,closed=true)=>HttpChannelOverHTTP@4a6678bf(exchange=null)[send=HttpSenderOverHTTP@f64ef97(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@3fd495df{s=START}],recv=HttpReceiverOverHTTP@3ea1630c(rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]
java.lang.NullPointerException
	at jnr.enxio.channels.KQSelector$Descriptor.access$000(KQSelector.java:75)
	at jnr.enxio.channels.KQSelector.handleChangedKey(KQSelector.java:247)
	at jnr.enxio.channels.KQSelector.interestOps(KQSelector.java:301)
	at jnr.enxio.channels.KQSelectionKey.interestOps(KQSelectionKey.java:60)
	at org.eclipse.jetty.io.ChannelEndPoint.updateKey(ChannelEndPoint.java:381)
	at org.eclipse.jetty.io.ChannelEndPoint$1.update(ChannelEndPoint.java:103)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processUpdates(ManagedSelector.java:358)
	at org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:329)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:357)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:181)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:132)
	at org.eclipse.jetty.io.ManagedSelector$$Lambda$7/1728790703.run(Unknown Source)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:760)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$1.run(QueuedThreadPool.java:678)
	at java.lang.Thread.run(Thread.java:745)
2018-05-31 21:58:19.304:INFO:oejs.AbstractConnector:main: Stopped UnixSocketConnector@de3a06f{HTTP/1.1,[http/1.1]}{/tmp/unix4830866990252150602.sock}

```

**Comment by joakime** [comment]
UnixSocketTest is fixed with commit fac63f4

**Comment by joakime** [comment]
@lachlan-roberts what does the error / stacktrace look like for the `SaveOptimizeTest` failure?

**Comment by lachlan-roberts** [comment]
@joakime the stacktrace was from the unix socket test, stacktrace is in the comment above

**Comment by lachlan-roberts** [comment]
@joakime it also seems like it would be more correct to `Assume.assumeTrue(OS.IS_UNIX);` in the `@before` section of the test rather than `assumeFalse(OS.IS_WINDOWS);` in each individual test, so should i keep these changes in

**Comment by joakime** [comment]
Don't worry about `@Before` in UnixSocketTest right now, the junit5 branch will fix that at the top level testclass itself later (not even running `@Before` and whatnot).

**Comment by joakime** [comment]
Note: the `OS` class in jetty-test-helper is deprecated and going away (in favor of the Junit5 OS features).

Also, I chose `!WINDOWS` to mimic the behavior with the jnr unixsocket implementation.

https://github.com/jnr/jnr-ffi/blob/8c6123604ed59b4bb52680e0a1c3d06f1327d7a4/src/main/java/jnr/ffi/Platform.java#L321-L323

``` java
    public final boolean isUnix() {
        return os != OS.WINDOWS;
    }
```

**Comment by lachlan-roberts** [comment]
@joakime okay I removed all of the changes to UnixSocketTest, now just has changes regarding the SaveOptimizeTest failures

**Comment by lachlan-roberts** [comment]
test is no longer failing on windows
https://jenkins.webtide.net/job/experiments/job/lachlan-test/4/testReport/

**Pull request merged**